keps/sig-network/20190423-service-lb-finalizer.md

```diff
@@ -0,0 +1,156 @@
+---
+title: Finalizer Protection for Service LoadBalancers
+authors:
+  - "@MrHohn"
+owning-sig: sig-network
+participating-sigs:
+  - sig-cloud-provider
+reviewers:
+  - "@andrewsykim"
+  - "@bowei"
+  - "@jhorwit2"
+  - "@jiatongw"
+approvers:
+  - "@andrewsykim"
+  - "@bowei"
+  - "@thockin"
+editor: TBD
+creation-date: 2019-04-23
+last-updated: 2019-04-23
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Finalizer Protection for Service LoadBalancers
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Release Signoff Checklist](#release-signoff-checklist)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+  - [Proposal](#proposal)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+We will be adding finalizer protection to ensure the Service resource is not
+fully deleted until the correlating load balancing resources are also deleted.
+Any service that has `type=LoadBalancer` (both existing and newly created ones)
+will be attached a service LoadBalancer finalizer, which should be removed by
+service controller upon the cleanup of related load balancing resources. Such
+finalizer protection mechanism will be released with phases to ensure downgrades
+can happen safely.
+
+## Motivation
+
+There are various cases where service controller can leave orphaned load
+balancing resources after services are deleted (ref discussion on
+https://github.com/kubernetes/kubernetes/issues/32157,
+https://github.com/kubernetes/kubernetes/issues/53451). We are periodically
+getting bug reports and customer issues that replicated such problem, which
+seems to be common enough and is worth to have a better mechanism for ensuring
+the cleanup of load balancing resources.
+
+### Goals
+
+Ensure the Service resource is not fully deleted until the correlating load
+balancing resources are also deleted.
+
+## Proposal
+
+We are going to define a finalizer for service LoadBalancers with name
+`kubernetes.io/service-load-balancer`. This finalizer will be attached to any
```

Comment:
`service.kubernetes.io/load-balancer-cleanup` ?   Do we have any convention for naming in-system finalizers?

Reply:
For volume we uses `kubernetes.io/pvc-protection` (ref https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/finalizer.go).

`load-balancer-cleanup` seems good though, as that is more clear what this is for.

Reply:
harumph.  Storage should have used a storage prefix.

Reply:
Agreed, updated to use `service.kubernetes.io/load-balancer-cleanup`.